source/MaterialXGenShader/ColorManagementSystem.cpp

```diff
@@ -88,4 +88,107 @@ ShaderNodePtr ColorManagementSystem::createNode(const ShaderGraph* parent, const
     return shaderNode;
 }
 
+
+void ColorManagementSystem::getPortConnections(ShaderGraph* graph, ShaderNode* colorTransformNode,
+                                               const TypeDesc* targetType, GenContext& context,
+                                               ShaderInput*& inputToConnect, ShaderOutput*& outputToConnect)
+{
+    const std::string colorTransformNodeName = colorTransformNode->getName();
+    ShaderOutput* colorTransformNodeOutput = colorTransformNode->getOutput(0);
+    ShaderInput* colorTransformNodeInput = colorTransformNode->getInput(0);
+
+    outputToConnect = colorTransformNodeOutput;
+    inputToConnect = colorTransformNodeInput;
+    if ((colorTransformNodeInput->getType() == Type::COLOR4 && targetType == Type::COLOR3) ||
+        (colorTransformNodeInput->getType() == Type::COLOR3 && targetType == Type::COLOR4))
+    {
+        const string convertToColor4String = "c3_to_" + colorTransformNodeName;
+        ShaderNode* convertToColor4Ptr = graph->getNode(convertToColor4String);
+        if (!convertToColor4Ptr)
+        {
+            NodeDefPtr toColor4NodeDef = _document->getNodeDef("ND_convert_color3_color4");
+            ShaderNodePtr convertToColor4 = ShaderNode::create(graph, convertToColor4String, *toColor4NodeDef, context);
+            graph->addNode(convertToColor4);
+            convertToColor4Ptr = convertToColor4.get();
+        }
+        const string convertToColor3String = colorTransformNodeName + "_to_c3";
+        ShaderNode* convertToColor3Ptr = graph->getNode(convertToColor3String);
+        if (!convertToColor3Ptr)
+        {
+            NodeDefPtr toColor3NodeDef = _document->getNodeDef("ND_convert_color4_color3");
+            ShaderNodePtr convertToColor3 = ShaderNode::create(graph, convertToColor3String, *toColor3NodeDef, context);
+            graph->addNode(convertToColor3);
+            convertToColor3Ptr = convertToColor3.get();
+        }
+
+        if (targetType == Type::COLOR3)
+        {
+            colorTransformNodeInput->makeConnection(convertToColor4Ptr->getOutput(0));
+            convertToColor3Ptr->getInput(0)->makeConnection(colorTransformNodeOutput);
+            inputToConnect = convertToColor4Ptr->getInput(0);
+            outputToConnect = convertToColor3Ptr->getOutput(0);
+        }
+        else
+        {
+            colorTransformNodeInput->makeConnection(convertToColor3Ptr->getOutput(0));
+            convertToColor4Ptr->getInput(0)->makeConnection(colorTransformNodeOutput);
+            inputToConnect = convertToColor3Ptr->getInput(0);
+            outputToConnect = convertToColor4Ptr->getOutput(0);
+        }
+    }
+}
+
+void ColorManagementSystem::connectNodeToShaderInput(ShaderGraph* graph, ShaderNodePtr colorTransformNodePtr, ShaderInput* shaderInput, GenContext& context)
```

Comment:
Move the input and output connection logic from shadergraph to the colorsystem.
Same logic as before, just uses a `getPortConnections` utility. If no conversion required then the logic is the same as what currently exists.

Reply:
Nitpick: For consistency maybe use the same signature as above and make the parameter `ShaderNode* colorTransformNode` instead of a shared pointer (and remove the conversion from shared to raw pointer below).

Reply:
I like the cleanup of moving color management node setup from shadergraph class to colorsystem class. Maybe take that all the way and move `ShaderGraph::addColorTransformNode` as well, to keep everything contained to colorsystem classes?

Reply:
I was thinking about this but thought it's a bit non explicit that adding a node + doing the connections.

Reply:
I'll fix the consistency issue.